### PR TITLE
Update CI to label-driven testing

### PR DESCRIPTION
**Pull request by Micky774**
# Description

With this PR, testing is mediated by labels such as: `ci-level {1, 2, 3}`. No tests will be run without at least one label present. At any given time, the highest-level label will take precedence, and only its corresponding level of tests will be run. The workflow triggers on labeling, (re-)opening PRs, and pushing commits. 

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .github/workflows/rocm-ci-lvl3.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+name: TransformerEngine CI Level 3
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-lvl3.yml]: Why this action is needed?

File: .github/workflows/rocm-ci.yml [hunk 2]
```diff
@@ -36,6 +52,7 @@ concurrency:
 
 jobs:
   build_and_test:
+    if: ${{ inputs.test_level == '3' || !(github.event_name == 'pull_request' && contains(github.event.pull_request.labels.*.name, 'Level 3')) }}
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci.yml]: What is the meaning of this condition: if Input.test_level==3 or NOT PR_with_level_3_label?

**Review by ipanfilo — COMMENTED** [review]
Maybe make approach more generic - ci_level% label specifies CI level to use.

File: .github/workflows/rocm-ci.yml (hunk 2 above)
Comment (Micky774) [reply .github/workflows/rocm-ci.yml]: We need the "NOT PR_with_level_3_label" or else when we have the level 3 label we necessarily run both level 1 and level 3. With that being said, we our level 3 workflow _calls this one_ so we need an exemption for that, which is provided by "Input.test_level==3"

File: .github/workflows/rocm-ci-lvl3.yml (hunk 1 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-lvl3.yml]: Overall to make it easier to guarantee that level 3 tests are run before PRs are merged in a visible way. I posted my thoughts on a more comprehensive refactoring of how we approach testing in our CI, but this is a standalone update.

File: .github/workflows/rocm-ci-dispatch.yml [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+            // Determine the current highest CI level from remaining labels
+            let level = '';
+            if (labels.includes('ci-level 3')) level = '3';
+            else if (labels.includes('ci-level 2')) level = '2';
+            else if (labels.includes('ci-level 1')) level = '1';
+
+            // Check if the removed level was higher than the current level
+            const currentLevel = level ? Number(level) : 0;
+            const removedWasHighest = removedLevel > 0 && removedLevel > currentLevel;
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: NIT: removedLevel > 0 is redundant

File: .github/workflows/rocm-ci-dispatch.yml [hunk 4]
```diff
@@ -0,0 +1,135 @@
+name: PR Automatic CI
+
+on:
+  pull_request:
+    branches:
+      - 'dev'
+      - 'release_v1.**_rocm'
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: NIT: release_v1 can be removed. We hardly support GHA for them now

File: .github/workflows/rocm-ci-dispatch.yml [hunk 5]
```diff
@@ -0,0 +1,135 @@
+name: PR Automatic CI
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml:5]: AMD copyright

File: .github/workflows/rocm-ci-dispatch.yml [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+              event: 'pull_request',
+              per_page: 100,
+            });
+
+            // Filter runs to find those that are for the same PR,
+            // are queued or in progress, and have a lower CI level
+            const toCancel = runs.data.workflow_runs.filter(run => {
+              const prMatch = (run.pull_requests || []).some(pr => pr.number === prNumber);
+              const active = run.status === 'queued' || run.status === 'in_progress';
+              if (action === 'synchronize') {
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: reopened action does not seem handled in this code

File: .github/workflows/rocm-ci-dispatch.yml [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+              workflow_id: 'rocm-ci-dispatch.yml',
+              event: 'pull_request',
+              per_page: 100,
+            });
+
+            // Filter runs to find those that are for the same PR,
+            // are queued or in progress, and have a lower CI level
+            const toCancel = runs.data.workflow_runs.filter(run => {
+              const prMatch = (run.pull_requests || []).some(pr => pr.number === prNumber);
+              const active = run.status === 'queued' || run.status === 'in_progress';
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: You can return false here if not prMatch or not active so do not need recheck it later. And, probably if run.id === currentRunId

File: .github/workflows/rocm-ci-dispatch.yml [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+                return prMatch && active && run.id !== currentRunId;
+              }
+              // For labeled/unlabeled actions, compare the CI levels
+              const title = run.display_title || '';
+              const match = title.match(/CI Level (\d+)/i);
+              const runLevel = match ? Number(match[1]) : 0;
+              // If a label was added, cancel runs with a lower level than the added label
+              if (action === 'labeled') {
+                const isLowerThanAdded = runLevel > 0 && runLevel < addedLevel;
+                return prMatch && active && isLowerThanAdded && run.id !== currentRunId;
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: can is LowerThanAdded be if run.id === currentRunId?

File: .github/workflows/rocm-ci-dispatch.yml [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      - name: Determine CI level from labels
+        id: set_level
+        uses: actions/github-script@v7
+        with:
+          script: |
+            const labels = (context.payload.pull_request.labels || [])
+              .map(label => label.name.toLowerCase());
+
+            // Determine if a CI level label was removed and what level it was   
+            const removedLabel = (context.payload.label?.name || '').toLowerCase();
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Why is it removedLabel? For labeled event it is addedLabel.
Maybe unlabeled events are not needed at all? If higher lablel is added - CI with new level is triggered. Otherwise the last CI run should still be relevant.

File: .github/workflows/rocm-ci.yml [hunk 10]
```diff
@@ -2,19 +2,30 @@
 #
 # See LICENSE for license information.
 
-name: TransformerEngine CI
+name: Test and Build Branch
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci.yml]: Build and Test?

File: .github/workflows/rocm-ci.yml [hunk 11]
```diff
@@ -31,12 +42,12 @@ on:
         default: false
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.ref }}
-  cancel-in-progress: true
+  group: ${{ github.event_name == 'workflow_call' && format('{0}-{1}-{2}', github.workflow, github.ref, github.run_id) || format('{0}-{1}', github.workflow, github.ref) }}
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci.yml]: Why modified group is needed for workflow_call ?

**Review by ipanfilo — CHANGES_REQUESTED** [review]

File: .github/workflows/rocm-ci-dispatch.yml (hunk 9 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: If a PR has labels `ci-level 1` and `ci-level 3` but then `ci-level 3` is removed, then we still need to run `ci-level 1` tests.

Comment (ipanfilo) [reply .github/workflows/rocm-ci-dispatch.yml]: > If a PR has labels `ci-level 1` and `ci-level 3` but then `ci-level 3` is removed, then we still need to run `ci-level 1` tests.

It ran Level 3 before, why does  it need run Level 1 now? The next time something is changed, yes, but why now?

Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Originally it was in case the highest-level test wasn't finished, in which case the commit still wouldn't necessarily have a passed test on the next-highest level yet. I've adjusted the script to explicitly check for tests of at least the new level, and skip dispatch if it exists.

File: .github/workflows/rocm-ci-dispatch.yml (hunk 5 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml:5]: Done

File: .github/workflows/rocm-ci-dispatch.yml (hunk 7 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Done

File: .github/workflows/rocm-ci-dispatch.yml (hunk 6 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Now handled as an early-skip

File: .github/workflows/rocm-ci.yml (hunk 11 above)
Comment (Micky774) [reply .github/workflows/rocm-ci.yml]: It's not needed, and now reverted

File: .github/workflows/rocm-ci.yml (hunk 10 above)
Comment (Micky774) [reply .github/workflows/rocm-ci.yml]: Fixed

File: .github/workflows/rocm-ci-dispatch.yml (hunk 4 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Done

File: .github/workflows/rocm-ci-dispatch.yml (hunk 3 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Done

File: .github/workflows/rocm-ci-dispatch.yml [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+            else if (labels.includes('ci-level 1')) level = '1';
+
+            // Check if the removed level was higher than the current level
+            const currentLevel = level ? Number(level) : 0;
+            const removedWasHighest = removedLevel > currentLevel;
+
+            let requires_dispatch = currentLevel > 0;
+            if (action === 'labeled') {
+              // Only dispatch when a CI-level label was added.
+              requires_dispatch &&= addedLevel > 0;
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: If it had ci-level 2 and then ci-level 1 is added, dispatching is is not needed so rather addedLevel >= currentLevel

File: .github/workflows/rocm-ci-dispatch.yml [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+            const labels = (context.payload.pull_request.labels || [])
+              .map(label => label.name.toLowerCase());
+            const action = context.payload.action;
+
+            // Determine if a CI level label was removed or added, and what level it was
+            const addedLevel = action === 'labeled' ? parseLevelLabel(context.payload.label?.name) : 0;
+            const removedLevel = action === 'unlabeled' ? parseLevelLabel(context.payload.label?.name) : 0;
+
+            // Determine the current highest CI level from remaining labels
+            let level = '';
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Nit: currentLevel can be assigned right in this block of code, w/o intermediate level

File: .github/workflows/rocm-ci-dispatch.yml [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+            const currentLevel = level ? Number(level) : 0;
+            const removedWasHighest = removedLevel > currentLevel;
+
+            let requires_dispatch = currentLevel > 0;
+            if (action === 'labeled') {
+              // Only dispatch when a CI-level label was added.
+              requires_dispatch &&= addedLevel > 0;
+            } else if (action === 'unlabeled') {
+              // Only dispatch downgrade when the removed label was the highest,
+              // and the now-highest level (if any) does not have a completed run.
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Why is it needed? If it running level 2 CI and ci-level 2 label is removed, why need to run level 1 CI?

File: .github/workflows/rocm-ci-dispatch.yml [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+              });
+
+              if (nextHighestAlreadyCompleted) {
+                requires_dispatch = false;
+              }
+            }
+            let test_level = requires_dispatch ? level : '';
+            core.setOutput('test_level', test_level);
+            core.setOutput('current_level', String(currentLevel));
+            const shouldCancelOthers = (test_level !== '') || (action === 'unlabeled' && removedWasHighest);
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Removing label should not automatically cancel running CI

Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Now only removing the last remaining ci-level label will cancel past CI. I want this to serve as an explicit request to save resources on unnecessary tests.

File: .github/workflows/rocm-ci-dispatch.yml (hunk 14 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: I've updated the logic to not dispatch/cancel if an equal or higher test run exists for the commit.

File: .github/workflows/rocm-ci-dispatch.yml (hunk 12 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Adjusted

File: .github/workflows/rocm-ci-dispatch.yml [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+# See LICENSE for license information.
+
+name: PR Automatic CI
+
+on:
+  pull_request:
+    branches:
+      - 'dev'
+      - 'release_v2.*_rocm'
+    types: [ labeled, unlabeled, synchronize, reopened ]
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Is unlabeled still used?

File: .github/workflows/rocm-ci-dispatch.yml [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+
+            const labels = (context.payload.pull_request.labels || [])
+              .map(label => parseLevelLabel(label.name));
+            const action = context.payload.action;
+            const currentLevel = labels.length ? Math.max(...labels) : 0;
+
+            // Determine if a CI level label was added, and what level it was.
+            const addedLevel = action === 'labeled' ? parseLevelLabel(context.payload.label?.name) : 0;
+
+            core.info(`Dispatch debug: action=${action}, addedLabel=${context.payload.label?.name || ''}, addedLevel=${addedLevel}, currentLevel=${currentLevel}, labels=[${labels.join(',')}]`);
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Info here makes sense only for labeled action

File: .github/workflows/rocm-ci-dispatch.yml [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+                });
+
+                core.info(`Dispatch debug: total workflow runs fetched=${runs.length}`);
+
+                const candidateRuns = runs.filter(run => {
+                  const prMatch = (run.pull_requests || []).some(pr => pr.number === prNumber);
+                  const sameSha = run.head_sha === headSha;
+                  const sameHeadRef = run.head_branch === context.payload.pull_request.head.ref;
+                  const activeOrDone =
+                    run.status === 'queued' ||
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: This looks like all possible status values

File: .github/workflows/rocm-ci-dispatch.yml [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+                  repo,
+                  workflow_id: 'rocm-ci-dispatch.yml',
+                  per_page: 100,
+                });
+
+                core.info(`Dispatch debug: total workflow runs fetched=${runs.length}`);
+
+                const candidateRuns = runs.filter(run => {
+                  const prMatch = (run.pull_requests || []).some(pr => pr.number === prNumber);
+                  const sameSha = run.head_sha === headSha;
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: head_sha is possible query parameter for listWorkflowRuns

File: .github/workflows/rocm-ci-dispatch.yml [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+                  owner,
+                  repo,
+                  workflow_id: 'rocm-ci-dispatch.yml',
+                  per_page: 100,
+                });
+
+                core.info(`Dispatch debug: total workflow runs fetched=${runs.length}`);
+
+                const candidateRuns = runs.filter(run => {
+                  const prMatch = (run.pull_requests || []).some(pr => pr.number === prNumber);
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: According to documentation this list contains all PR's with matching head_sha or head_branch so it does not help to associate the PR. IMHO the most precise matching is branch query parameter to match the PR target branch and head_sha and head_branch for the run both match the PR too.
Even this one is not ideal because PR run does merging of source and target so strictly speaking it may run different code after each target update

File: .github/workflows/rocm-ci-dispatch.yml [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+                  // Inspect called-workflow job names, which include level.
+                  const jobsResp = await github.rest.actions.listJobsForWorkflowRun({
+                    owner,
+                    repo,
+                    run_id: run.id,
+                    per_page: 100,
+                  });
+
+                  for (const job of jobsResp.data.jobs || []) {
+                    const m = (job.name || '').match(/(?:CI\s+)?Level\s*(\d+)/i);
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Can jobs in one run have different levels?

File: .github/workflows/rocm-ci-dispatch.yml [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+                    if (m) {
+                      detectedLevel = Math.max(detectedLevel, Number(m[1]));
+                    }
+                    core.info(`Dispatch debug: run_id=${run.id}, job_id=${job.id}, job_name=${job.name}, detectedLevel=${detectedLevel}`);
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Potentially excessive logging with this and the next message in the loop

File: .github/workflows/rocm-ci-dispatch.yml [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+                core.info(`Dispatch debug: post-existing-run check requiresDispatch=${requiresDispatch}`);
+              }
+            } else if (action === 'synchronize' || action === 'reopened') {
+              requiresDispatch = currentLevel > 0;
+            }
+            core.info(`Dispatch debug: final requiresDispatch=${requiresDispatch}, output test_level=${requiresDispatch ? String(currentLevel) : ''}`);
+            core.setOutput('test_level', requiresDispatch ? String(currentLevel) : '');
+
+  maybe_cancel_all:
+    if: ${{ github.event.action == 'unlabeled' }}
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: I'm still opposed of using labels as a tool to stop workflowd. User may cancel workflow in actions menu, if they want

File: .github/workflows/rocm-ci-dispatch.yml (hunk 20 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Could you help me understand what scenario might this fail in specifically? I've updated it to only rely on the PR match + head sha. Is there a situation where that wouldn't be precise enough?

File: .github/workflows/rocm-ci-dispatch.yml (hunk 19 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Updated

File: .github/workflows/rocm-ci-dispatch.yml (hunk 18 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: I misunderstood the API a bit -- I've updated to what I think is a more accurate reflection of what I want (basically it can be any of the active statuses, OR it can be completed AND successful).

File: .github/workflows/rocm-ci-dispatch.yml (hunk 17 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Updated

File: .github/workflows/rocm-ci-dispatch.yml (hunk 21 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: No, simplified and updated.

File: .github/workflows/rocm-ci-dispatch.yml (hunk 22 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: Updated

File: .github/workflows/rocm-ci-dispatch.yml (hunk 23 above)
Comment (Micky774) [reply .github/workflows/rocm-ci-dispatch.yml]: I've removed the cancellation now.

File: .github/workflows/rocm-ci-dispatch.yml (hunk 20 above)
Comment (ipanfilo) [reply .github/workflows/rocm-ci-dispatch.yml]: One scenario is base branch change for PR. And the other is target branch update - they do not directly affect PR code, so head_sha remains the same but test results may differ.
It is OK not to automatically rerun CI in those cases though because they are rather corner cases and mainstream usage will not include label playing w/o code touching.
By the same reason it is not to worry mach about multiple PRs from the same commit to different targets.

File: .github/workflows/rocm-ci-dispatch.yml [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+                const candidateRuns = runs.filter(run => {
+                  const prMatch = (run.pull_requests || []).some(pr => pr.number === prNumber);
+                  const completedNotSuccess = run.status === 'completed' && run.conclusion !== 'success';
+                  return prMatch && !completedNotSuccess;
+                });
+
+                core.info(`Dispatch debug: candidate runs matching PR+sha=${candidateRuns.length}`);
+
+                let satisfiesAddedLevel = false;
+                for (const run of candidateRuns) {
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci-dispatch.yml]: Doesn't run-name (CI Level) appear in github.rest.actions.listWorkflowRuns?

**Review by ipanfilo — APPROVED** [review]
